crunchers/cruncher_top_cards.h

```diff
@@ -0,0 +1,250 @@
+  CURRENT_FIELD(tfu_count, uint64_t, 0);
+  CURRENT_FIELD(fav_count, uint64_t, 0);
+  CURRENT_FIELD(unfav_count, uint64_t, 0);
+  CURRENT_FIELD(seen_count, uint64_t, 0);
+
+  bool Empty() {
+    return ctfo_count == 0 && skip_count == 0 && tfu_count == 0 && fav_count == 0 && unfav_count == 0 &&
+           seen_count == 0;
+  }
+};
+
+using TopCardsCruncherResponse = std::vector<TopCardsCruncherResponseItem>;
+
+template <typename NAMESPACE>
+struct TopCardsCruncherImpl {
+  using Transaction_Z = typename NAMESPACE::Transaction_T9220981828355492272;
+  using EventLogEntry = typename NAMESPACE::EventLogEntry;
+  using CardDeleted = typename NAMESPACE::Persisted_CardDeleted;
+  using iOSGenericEvent = typename NAMESPACE::iOSGenericEvent;
+  using CID = typename NAMESPACE::CID;
+  using event_t = typename NAMESPACE::CTFOLogEntry;
+  using card_t = TopCardsCruncherResponseItem;
+  using value_t = std::vector<card_t>;
+
+  TopCardsCruncherImpl(const TopCardsCruncherArgs& args) : args_(args) {}
+  virtual ~TopCardsCruncherImpl() = default;
+
+  void OnEvent(const event_t& e, idxts_t idxts) {
+    current_us_ = idxts.us;
+    if (Exists<Transaction_Z>(e)) {
+      const auto& transaction = Value<Transaction_Z>(e);
+      for (const auto& mutation : transaction.mutations) {
+        if (Exists<CardDeleted>(mutation)) {
+          OnCardDeleted(Value<CardDeleted>(mutation));
+        }
+      }
+    } else if (Exists<EventLogEntry>(e)) {
+      OnEventLogEntry(Value<EventLogEntry>(e));
+    }
+    while (!events_list_.empty() && events_list_.back().us + args_.interval <= current_us_) {
+      TimeWindowLeft(events_list_.back());
+      events_list_.pop_back();
+    }
+  }
+
+  value_t GetValue(size_t n = 0) const {
+    if (!n) n = args_.top_size;
+    value_t result;
+    result.reserve(n);
+    for (const auto& bucket : top_cards_map_) {
+      for (const auto cid : bucket.second) {
+        const auto cit = cards_map_.find(cid);
+        result.push_back(cit->second);
+        if (result.size() >= n) {
+          return result;
+        }
+      }
+    }
+    return result;
+  }
+
+ private:
+  enum class CTFO_EVENT : int { SEEN, SKIP, CTFO, TFU, FAV_CARD, UNFAV_CARD };
+
+  struct CardEvent final {
+    CID cid;
+    CTFO_EVENT type;
+    std::chrono::microseconds us;
+  };
+  using event_list_t = std::deque<CardEvent>;
+  using cards_map_t = std::unordered_map<CID, card_t, current::CurrentHashFunction<CID>>;
+  using top_cards_map_t =
+      std::map<int64_t, std::unordered_set<CID, current::CurrentHashFunction<CID>>, std::greater<uint64_t>>;
+
+  void TimeWindowsEnter(CardEvent&& e) {
+    const auto cit = cards_map_.find(e.cid);
+    if (cit != cards_map_.end()) {
+      card_t& card = cit->second;
+      auto it = top_cards_map_.find(card.rate);
+      it->second.erase(e.cid);
+      ApplyCardEvent(card, e.type, false /*rollback*/);
+      top_cards_map_[card.rate].insert(e.cid);
+      if (it->second.empty()) top_cards_map_.erase(it);
```

Comment:
Moving this statement to right after `it->second.erase(e.cid)` is probably a bit more readable.